File: Serenity.Script.UI/Grid/EntityGrid.cs
using jQueryApi;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [IncludeGenericArguments(false), ScriptName("EntityGrid")]
    public abstract class EntityGrid<TEntity, TOptions> : DataGrid<TEntity, TOptions>
        where TEntity: class, new()
        where TOptions: class, new()
    {
        private string entityType;
        private string displayName;
        private string itemName;
        private string service;
        private Type dialogType;

        public EntityGrid(jQueryObject container, TOptions opt = null)
            : base(container, opt)
        {
        }

        protected override bool UsePager()
        {
            return true;
        }

        protected override void CreateToolbarExtensions()
        {
            CreateIncludeDeletedButton();
            CreateQuickSearchInput();
        }

        protected override string GetInitialTitle()
        {
            return GetDisplayName();
        }

        protected override string GetLocalTextPrefix()
        {
            var result = base.GetLocalTextPrefix();

            if (result.IsEmptyOrNull())
                return GetEntityType();

            return result;
        }

        protected virtual string GetEntityType()
        {
            if (entityType == null)
            {
                var typeAttributes = this.GetType().GetCustomAttributes(typeof(EntityTypeAttribute), true);
                if (typeAttributes.Length == 1)
                {
                    entityType = typeAttributes[0].As<EntityTypeAttribute>().Value;
                    return entityType;
                }

                // typeof(TEntity).Name'i kullanamayız, TEntity genelde Serializable ve Imported olduğundan dolayı tipi Object e karşılık geliyor!

                // remove global namespace
                var name = this.GetType().FullName;
                var px = name.IndexOf(".");
                if (px >= 0)
                    name = name.Substring(px + 1);

                if (name.EndsWith("Grid"))
                    name = name.Substr(0, name.Length - 4);
                else if (name.EndsWith("SubGrid"))
                    name = name.Substr(0, name.Length - 7);

                entityType = name;
            }

            return entityType;
        }

        protected virtual string GetDisplayName()
        {
            if (displayName == null)
            {
                var attributes = this.GetType().GetCustomAttributes(typeof(DisplayNameAttribute), true);
                if (attributes.Length >= 1)
                {
                    displayName = attributes[0].As<DisplayNameAttribute>().DisplayName;
                    displayName = LocalText.GetDefault(displayName, displayName);
                }
                else
                    displayName = (Q.TryGetText(GetLocalTextDbPrefix() + "EntityPlural") ?? GetEntityType());
            }

            return displayName;
        }

        protected virtual string GetItemName()
        {
            if (itemName == null)
            {
                var attributes = this.GetType().GetCustomAttributes(typeof(ItemNameAttribute), true);
                if (attributes.Length >= 1)
                {
                    itemName = attributes[0].As<ItemNameAttribute>().Value;
                    itemName = LocalText.GetDefault(itemName, itemName);
                }
                else
                    itemName = (Q.TryGetText(GetLocalTextDbPrefix() + "EntitySingular") ?? GetEntityType());
            }

            return itemName;
        }

        protected override string GetAddButtonCaption()
        {
            return String.Format(Q.Text("Controls.EntityGrid.NewButton"), GetItemName());
        }

        protected override List<ToolButton> GetButtons()
        {
            var self = this;
            var buttons = new List<ToolButton>();

            buttons.Add(new ToolButton
            {
                Title = GetAddButtonCaption(),
                CssClass = "add-button",
                Hotkey = "alt+n",
                OnClick = delegate {
                    self.AddButtonClick();
                }
            });

            buttons.Add(NewRefreshButton(noText: true));

            return buttons;
        }

        protected ToolButton NewRefreshButton(bool noText = false)
        {
            var self = this;

            return new ToolButton
            {
                Title = noText ? null : Q.Text("Controls.EntityGrid.RefreshButton"),
                Hint = noText ? Q.Text("Controls.EntityGrid.RefreshButton") : null,
                CssClass = "refresh-button",
                OnClick = delegate
                {
                    self.Refresh();
                }
            };
        }

        protected virtual void AddButtonClick()
        {
            EditItem(new object().As<TEntity>());
        }

        protected override void EditItem(object entityOrId)
        {
            CreateEntityDialog(GetItemType(), dlg =>
            {
                var dialog = dlg as IEditDialog;
                if (dialog != null)
                {
                    dialog.Load(entityOrId, () => dialog.DialogOpen(), null);
                    return;
                }

                throw new InvalidOperationException(String.Format(
                    "{0} doesn't implement IEditDialog!", dlg.GetType().FullName));
            });
        }

        protected override void EditItemOfType(string itemType, object entityOrId)
        {
            if (itemType == GetItemType())
            {
                EditItem(entityOrId);
                return;
            }

            CreateEntityDialog(itemType, dlg =>
            {
                var dialog = dlg as IEditDialog;
                if (dialog != null)
                {
                    dialog.Load(entityOrId, () => dialog.DialogOpen(), null);
                    return;
                }

                throw new InvalidOperationException(String.Format(
                    "{0} doesn't implement IEditDialog!", dlg.GetType().FullName));
            });
        }

        protected virtual string GetService()
        {
            if (service == null)
            {
                var attributes = this.GetType().GetCustomAttributes(typeof(ServiceAttribute), true);
                if (attributes.Length >= 1)
                    service = attributes[0].As<ServiceAttribute>().Value;
                else
                    service = this.GetEntityType().Replace('.', '/');
            }

            return service;
        }

        protected override SlickRemoteViewOptions GetViewOptions()
        {
            var opt = base.GetViewOptions();
            opt.Url = Q.ResolveUrl("~/Services/" + GetService() + "/List");
            return opt;
        }

        protected override string GetItemType()
        {
            return GetEntityType();
        }

        [Obsolete("Use InitDialog"), InlineCode("InitDialog({dialog})")]
        protected void InitEntityDialog(Widget dialog)
        {
            InitDialog(dialog);
        }

        protected virtual void InitDialog(Widget dialog)
        {
            var self = this;
            dialog.BindToDataChange(this, (e, dci) => self.SubDialogDataChange());
        }

        protected virtual void InitEntityDialog(string itemType, Widget dialog)
        {
            if (itemType == GetItemType())
            {
                InitDialog(dialog);
                return;
            }

            var self = this;
            dialog.BindToDataChange(this, (e, dci) => self.SubDialogDataChange());
        }

        protected virtual Widget CreateEntityDialog(string itemType, Action<Widget> callback)
        {
            var dialogClass = GetDialogTypeFor(itemType);

            var dialog = Serenity.Widget.CreateOfType(
                widgetType: dialogClass,
                options: (object)GetDialogOptionsFor(itemType),
                init: (d) =>
                {
                    InitEntityDialog(itemType, d);
                    if (callback != null)
                        callback(d);
                });

            return dialog;
        }

        protected virtual dynamic GetDialogOptions()
        {
            return new { };
        }

        protected virtual dynamic GetDialogOptionsFor(string itemType)
        {
            if (itemType == GetItemType())
                return GetDialogOptions();

            return new { };
        }

        protected virtual Type GetDialogTypeFor(string itemType)
        {
            if (itemType == GetItemType())
                return GetDialogType();

            return DialogTypeRegistry.Get(itemType);
        }

        protected virtual Type GetDialogType()
        {
            if (dialogType == null)
            {
                var attributes = this.GetType().GetCustomAttributes(typeof(DialogTypeAttribute), true);
                if (attributes.Length >= 1)
                    dialogType = attributes[0].As<DialogTypeAttribute>().Value;
                else
                {
                    dialogType = DialogTypeRegistry.Get(GetEntityType());
                }
            }

            return dialogType;
        }
    }

    [Imported, IncludeGenericArguments(false), ScriptName("EntityGrid")]
    public abstract class EntityGrid<TEntity> : EntityGrid<TEntity, object>
        where TEntity : class, new()
    {
        public EntityGrid(jQueryObject container)
            : base(container)
        {
        }
    }
}

File: src/Serenity.Scripts/CoreLib/UI/DataGrid/EntityGrid.ts
import { Decorators, DialogTypeAttribute, DisplayNameAttribute, EntityTypeAttribute, ItemNameAttribute, ServiceAttribute } from "../../Decorators";
import { IEditDialog } from "../../Interfaces";
import { Authorization, endsWith, format, getInstanceType, getTypeFullName, HandleRouteEventArgs, isEmptyOrNull, LT, replaceAll, resolveUrl, Router, safeCast, text, tryGetText } from "../../Q";
import { DialogTypeRegistry } from "../../Types/DialogTypeRegistry";
import { EditorUtils } from "../Editors/EditorUtils";
import { SubDialogHelper } from "../Helpers/SubDialogHelper";
import { ToolButton } from "../Widgets/Toolbar";
import { Widget, WidgetDialogClass } from "../Widgets/Widget";
import { ColumnPickerDialog } from "./ColumnPickerDialog";
import { DataGrid } from "./DataGrid";

@Decorators.registerClass('Serenity.EntityGrid')
export class EntityGrid<TItem, TOptions> extends DataGrid<TItem, TOptions> {

    constructor(container: JQuery, options?: TOptions) {
        super(container, options);

        this.element.addClass('route-handler')
            .on('handleroute.' + this.uniqueName, (_, args: any) => this.handleRoute(args));
    }

    protected handleRoute(args: HandleRouteEventArgs): void {
        if (!!args.handled)
            return;

        if (!!(args.route === 'new')) {
            args.handled = true;
            this.addButtonClick();
            return;
        }

        var oldRequests = jQuery["active"];

        var parts = args.route.split('/');
        if (!!(parts.length === 2 && parts[0] === 'edit')) {
            args.handled = true;
            this.editItem(decodeURIComponent(parts[1]));
        }
        else if (!!(parts.length === 2 && parts[1] === 'new')) {
            args.handled = true;
            this.editItemOfType(parts[0], null);
        }
        else if (!!(parts.length === 3 && parts[1] === 'edit')) {
            args.handled = true;
            this.editItemOfType(parts[0], decodeURIComponent(parts[2]));
        }
        else
            return;

        if (jQuery["active"] > oldRequests && args.handled && args.index >= 0 && args.index < args.parts.length - 1) {
            $(document).one('ajaxStop', () => {
                setTimeout(() => Router.resolve('#' + args.parts.join('/+/')), 1);
            });
        }
    }

    protected usePager(): boolean {
        return true;
    }

    protected createToolbarExtensions(): void {
        this.createIncludeDeletedButton();
        this.createQuickSearchInput();
    }

    protected getInitialTitle(): string {
        return this.getDisplayName();
    }

    protected getLocalTextPrefix(): string {
        var result = super.getLocalTextPrefix();

        if (isEmptyOrNull(result)) {
            return this.getEntityType();
        }

        return result;
    }

    private entityType: string;

    protected getEntityType(): string {
        if (this.entityType != null)
            return this.entityType;

        var attr = this.attrs(EntityTypeAttribute);

        if (attr.length === 1) {
            return (this.entityType = attr[0].value);
        }

        var name = getTypeFullName(getInstanceType(this));

        var px = name.indexOf('.');
        if (px >= 0) {
            name = name.substring(px + 1);
        }

        if (endsWith(name, 'Grid')) {
            name = name.substr(0, name.length - 4);
        }
        else if (endsWith(name, 'SubGrid')) {
            name = name.substr(0, name.length - 7);
        }

        this.entityType = name;

        return this.entityType;
    }

    private displayName: string;

    protected getDisplayName(): string {
        if (this.displayName != null)
            return this.displayName;

        var attr = this.attrs(DisplayNameAttribute);
        if (attr.length >= 1) {
            this.displayName = attr[0].displayName;
            this.displayName = LT.getDefault(this.displayName, this.displayName);
        }
        else {
            this.displayName = tryGetText(this.getLocalTextDbPrefix() + 'EntityPlural');
            if (this.displayName == null)
                this.displayName = this.getEntityType();
        }

        return this.displayName;
    }

    private itemName: string;

    protected getItemName(): string {
        if (this.itemName != null)
            return this.itemName;

        var attr = this.attrs(ItemNameAttribute);
        if (attr.length >= 1) {
            this.itemName = attr[0].value;
            this.itemName = LT.getDefault(this.itemName, this.itemName);
        }
        else {
            this.itemName = tryGetText(this.getLocalTextDbPrefix() + 'EntitySingular');
            if (this.itemName == null)
                this.itemName = this.getEntityType();
        }

        return this.itemName;
    }

    protected getAddButtonCaption(): string {
        return format(text('Controls.EntityGrid.NewButton'), this.getItemName());
    }

    protected getButtons(): ToolButton[] {

        var buttons: ToolButton[] = [];
        buttons.push({
            title: this.getAddButtonCaption(),
            cssClass: 'add-button',
            hotkey: 'alt+n',
            onClick: () => {
                this.addButtonClick();
            },
            disabled: () => !this.hasInsertPermission() || this.readOnly
            });

        buttons.push(this.newRefreshButton(true));
        buttons.push(ColumnPickerDialog.createToolButton(this as any));

        return buttons;
    }

    protected newRefreshButton(noText?: boolean): ToolButton {
        return {
            title: (noText ? null : text('Controls.EntityGrid.RefreshButton')),
            hint: (noText ? text('Controls.EntityGrid.RefreshButton') : null),
            cssClass: 'refresh-button',
            onClick: () => {
                this.refresh();
            }
        };
    }

    protected addButtonClick(): void {
        this.editItem(new Object());
    }

    protected editItem(entityOrId: any): void {
        this.createEntityDialog(this.getItemType(), dlg => {
            var dialog = safeCast(dlg, IEditDialog);
            if (dialog != null) {
                dialog.load(entityOrId, () => {
                    dialog.dialogOpen(this.openDialogsAsPanel);
                });

                return;
            }

            throw new Error(
                format("{0} doesn't implement IEditDialog!",
                    getTypeFullName(getInstanceType(dlg))));
        });
    }

    protected editItemOfType(itemType: string, entityOrId: any): void {

        if (itemType === this.getItemType()) {
            this.editItem(entityOrId);
            return;
        }

        this.createEntityDialog(itemType, dlg => {
            var dialog = safeCast(dlg, IEditDialog);
            if (dialog != null) {
                dialog.load(entityOrId, () =>
                    dialog.dialogOpen(this.openDialogsAsPanel));
                return;
            }

            throw new Error(
                format("{0} doesn't implement IEditDialog!",
                    getTypeFullName(getInstanceType(dlg))));
        });
    }

    private service: string;

    protected getService(): string {
        if (this.service != null)
            return this.service;

        var attr = this.attrs(ServiceAttribute);
        if (attr.length >= 1)
            this.service = attr[0].value;
        else
            this.service = replaceAll(this.getEntityType(), '.', '/');

        return this.service;
    }

    protected getViewOptions(): Slick.RemoteViewOptions {
        var opt = super.getViewOptions();
        opt.url = resolveUrl('~/Services/' + this.getService() + '/List');
        return opt;
    }

    protected getItemType() {
        return this.getEntityType();
    }

    protected routeDialog(itemType: string, dialog: Widget<any>) {
        Router && Router.dialog && Router.dialog(this.element, dialog.element, () => {
            var hash = '';

            if (itemType !== this.getItemType())
                hash = itemType + '/';

            if (!!(dialog != null && (dialog as any).entityId != null))
                hash += 'edit/' + (dialog as any).entityId.toString();
            else
                hash += 'new';

            return hash;
        });
    }

    protected getInsertPermission(): string {
        return null;
    }

    protected hasInsertPermission(): boolean {
        var insertPermission = this.getInsertPermission();
        return insertPermission == null || Authorization.hasPermission(this.getInsertPermission());
    }

    protected transferDialogReadOnly(dialog: Widget<any>) {
        if (this.readOnly)
            EditorUtils.setReadOnly(dialog, true);
    }

    protected initDialog(dialog: Widget<any>): void {
        SubDialogHelper.bindToDataChange(dialog, this, (e, dci) => {
            this.subDialogDataChange();
        }, true);

        this.transferDialogReadOnly(dialog);
        this.routeDialog(this.getItemType(), dialog);
    }

    protected initEntityDialog(itemType: string, dialog: Widget<any>): void {
        if (itemType === this.getItemType()) {
            this.initDialog(dialog);
            return;
        }

        SubDialogHelper.bindToDataChange(dialog, this, (e, dci) => {
            this.subDialogDataChange();
        }, true);

        this.transferDialogReadOnly(dialog);
        this.routeDialog(itemType, dialog);
    }

    protected createEntityDialog(itemType: string, callback?: (dlg: Widget<any>) => void): Widget<any> {
        var dialogClass = this.getDialogTypeFor(itemType);
        var dialog = Widget.create({
            type: dialogClass,
            options: this.getDialogOptionsFor(itemType),
            init: d => {
                this.initEntityDialog(itemType, d);
                callback && callback(d);
            }
        });

        return dialog;
    }

    protected getDialogOptions(): JQueryUI.DialogOptions {
        return {};
    }

    protected getDialogOptionsFor(itemType: string): JQueryUI.DialogOptions {
        if (itemType === this.getItemType())
            return this.getDialogOptions();

        return {};
    }

    protected getDialogTypeFor(itemType: string): { new(...args: any[]): Widget<any> } {

        if (itemType === this.getItemType()) {
            return this.getDialogType();
        }

        return DialogTypeRegistry.get(itemType) as any;
    }

    private dialogType: WidgetDialogClass;

    protected getDialogType(): { new(...args: any[]): Widget<any> } {

        if (this.dialogType != null)
            return this.dialogType;

        var attr = this.attrs(DialogTypeAttribute);
        if (attr.length >= 1)
            this.dialogType = attr[0].value;
        else
            this.dialogType = DialogTypeRegistry.get(this.getEntityType());

        return this.dialogType;
    }
}
